pkg/storagev2/leveldb/store.go

```diff
@@ -0,0 +1,292 @@
+	path    string
+	closeLk sync.RWMutex
+}
+
+var _ storageV2.Store = (*Store)(nil)
+var _ storageV2.Tx = (*transaction)(nil)
+
+// NewDatastore returns a new datastore backed by leveldb
+// for path == "", an in memory backend will be chosen (TODO)
+func NewLevelDBStore(path string, opts *opt.Options) (storageV2.Store, error) {
+	var err error
+	var db *leveldb.DB
+
+	if path == "" {
+		db, err = leveldb.Open(storage.NewMemStorage(), opts)
+	} else {
+		db, err = leveldb.OpenFile(path, opts)
+		if errors.IsCorrupted(err) && !opts.GetReadOnly() {
+			db, err = leveldb.RecoverFile(path, opts)
+		}
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	ds := Store{
+		DB:   db,
+		path: path,
+	}
+	return &ds, nil
+}
+
+func (s *Store) Count(key storageV2.Key) (c int, err error) {
+	keys := util.BytesPrefix([]byte(key.Namespace() + "/"))
+
+	iter := s.DB.NewIterator(keys, nil)
+	defer iter.Release()
+
+	for iter.Next() {
+		c++
+	}
+
+	return c, iter.Error()
+}
+
+func (s *Store) Put(item storageV2.Item) error {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	value, err := item.Marshal()
+	if err != nil {
+		return fmt.Errorf("failed serializing %w", err)
+	}
+
+	key := []byte(item.Namespace() + "/" + item.ID())
+
+	return s.DB.Put(key, value, &opt.WriteOptions{Sync: true})
+}
+
+func (s *Store) Get(item storageV2.Item) error {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(item.Namespace() + "/" + item.ID())
+
+	val, err := s.DB.Get(key, nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return storageV2.ErrNotFound
+		}
+		return err
+	}
+
+	if err = item.Unmarshal(val); err != nil {
+		return fmt.Errorf("failed decoding value %w", err)
+	}
+
+	return nil
+}
+
+func (s *Store) Has(sKey storageV2.Key) (exists bool, err error) {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(sKey.Namespace() + "/" + sKey.ID())
+
+	return s.DB.Has(key, nil)
+}
+
+func (s *Store) GetSize(sKey storageV2.Key) (size int, err error) {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(sKey.Namespace() + "/" + sKey.ID())
+
+	val, err := s.DB.Get(key, nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return 0, storageV2.ErrNotFound
+		}
+		return 0, err
+	}
+
+	return len(val), nil
+}
+
+func (s *Store) Delete(item storageV2.Key) (err error) {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(item.Namespace() + "/" + item.ID())
+
+	return s.DB.Delete(key, &opt.WriteOptions{Sync: true})
+}
+
+func (s *Store) Iterate(q storageV2.Query, fn storageV2.IterateFn) error {
+	if err := q.Validate(); err != nil {
+		return fmt.Errorf("failed iteration: %w", err)
+	}
+
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	var retErr *multierror.Error
+
+	namespace := q.Factory().Namespace()
+	prefix := namespace + "/"
+
+	keys := util.BytesPrefix([]byte(prefix))
+
+	iter := s.DB.NewIterator(keys, nil)
+
+	switch q.Order {
+	case storageV2.KeyAscendingOrder:
+		for iter.Next() {
```

Comment:
So I think this can be simplified. Both the `Prev` and `Next` functions have same signature. So we can have just 1 for loop basically.
```
nextF := iter.Next

if q.Order == KeyOrderDescending {
    nextF = func() bool {
        nextF = iter.Prev
        return iter.Last()
   }
}
```

Reply:
I think this way we dont need a separate function for `apply` maybe.